File: noaa-requests/NOAA_Water_Level_Downloader/WL_downloader_V2.m
%{
Data download script

IMPORTANT: IF STATION IS NOT PRESENT ON DATA STRUCTURE StationList

Prodcut download limit per request
Predictions:
31 days (no matter the interval of the product)
Water Level Measurements:
31 days (6-minute product)
365 days (hourly product)
none (monthly or peaks data)

Flag Legend:
flag1 - defines the WL product that is selected ( hourly - '1', 6 min - '6', monthly - 'm' )
flag2 - indicates if station is in the great lakes hence there are no tidal predictions.
 (1 - station is on great lakes, 0 - stations is not on great lakes)
flag3 - indicates if tidal predictions product is high/low (bad) or not (0 -
prefered data interval was found)
-----------------------------------------------------------------------------------------------
EDIT HISTORY:
User        Date        Description
FGM      25/09/2019     Fixed filtering issues discarding tidal stations
FGM      26/10/2019     Added the functionality of donwloading monthly products
FGM       3/10/2019     Fixed filtering issues discarding tidal stations
FGM       5/10/2019     Added record length calculator for downloaded product and NaN gap filler
FGM       7/10/2019     Added the capacity of downloading multiple products at once
FGM       9/10/2019     Tested and vaidated new functionalities
FGM       16/10/2019    Added try/catch statement to datum_selector.m (Great Lakes case)
FGM       16/10/2019    Added if statement to prediction_interval_selector.m (Great Lakes case)
FGM       16/10/2019    Added if statement for vector length check
FGM       20/10/2019    Added failsafe to WL_measurements_product_selector_V2. (flag4)
%}
function [Sdata,notFound] = WL_downloader_V2(idList,StationList,requested_datum,prod,opMode,dBeg,dEnd) %% ADD PATHS
    if ~ismember(opMode,1:2)
        error('Please use a valid operational mode (1 - full record, 2 - Specific Date)');
    end
    %% START TIMER
    timing2(1).t1 = datetime('now');
    
    
    %% ADD PATHS
    addpath(genpath('NOAA_Download_Functions/'));
    
    %% DEFINE INITIAL VARIABLES
    % Prefered Datum for Data Download
    datum = 'MSL';
    % Prefered Time Zone for Data Download
    timezone = 'GMT';
    % Prefered Units for Data Download
    units = 'metric';
    % Download Data Format
    format = 'csv';
    % NOAA CO-OPS URL
    gen_url = 'https://api.tidesandcurrents.noaa.gov';
    % 5 minute timeout for webread
    options = weboptions('Timeout',800);
    
    %% Check if stations are in inventory
    % Compute Logical Matrix (Columns Correspond To Cases In want)
    logical_matrix = cell2mat(cellfun(@(c)strcmp(c,{StationList.id}'),idList,'UniformOutput',false));
    % Find Matching Stations
    [mx,mindx] = max(logical_matrix,[],1);
    % Single out stations that were not found in StationList
    notFound = idList(mx==0);
    % Create valid station id list
    idList =  {StationList(mindx(mx==1)).id};
    
    %% INITIALIZE COUNTER
    ctr=1;
    
    %% DOWNLOAD DATA FOR EACH STATION
    for ii = 1:length(idList)
        ['------------------ ',num2str(ii),' / ',num2str(length(idList)),' ---------------------']
        for pp = 1:length(prod)
            %% EXTRACT STATION FROM DATA STRUCTURE
            % Get Station ID
            station = idList{ii};
            % Get All Fields for Station
            dStruct = StationList(strcmp(station,{StationList.id}')==1);
            % Assign relevant data to new data structure
            Sdata(ctr).id = station;
            Sdata(ctr).name = dStruct.name;
            Sdata(ctr).lon = dStruct.lon;
            Sdata(ctr).lat = dStruct.lat;
            Sdata(ctr).state = dStruct.state;
            
            %% CHECK WL MEASUREMENTS PRODUCTS AVAILABLE
            [flag4,flag1,indx] = WL_measurements_product_selector_V2(dStruct,prod(pp));
            %     [flag1,indx] = WL_measurements_product_selector(dStruct);
            if flag4==1
                Sdata(ctr).WL_datum = 'Not found';
                Sdata(ctr).TP_datum = 'Not found';
                Sdata(ctr).WL_downloaded_product = 'Not found';
                Sdata(ctr).TP_downloaded_product = 'Not found';
                Sdata(ctr).record_length = 'Not found';
                Sdata(ctr).WL = 'Not found';
                Sdata(ctr).TP = 'Not found';
                ctr=ctr+1;
            else
                %% CHECK IF DATE RANGE OF INTEREST IS AVAILABLE
                if opMode == 2
                    [indx, dEnd, dBeg, dummy3] = date_search(dStruct,dBeg,dEnd,indx);
                    if isempty(dummy3)
                        dStruct.startDate(indx) = {[datestr(dBeg,'yyyy-mm-dd HH:MM:SS'),' GMT']};
                        dStruct.endDate(indx) = {[datestr(dEnd,'yyyy-mm-dd HH:MM:SS'),' GMT']};
                    else
                        Sdata(ctr).WL_datum = 'Not found';
                        Sdata(ctr).TP_datum = 'Not found';
                        Sdata(ctr).WL_downloaded_product = 'Not found';
                        Sdata(ctr).TP_downloaded_product = 'Not found';
                        Sdata(ctr).record_length = 'Not found';
                        Sdata(ctr).WL = 'Not found';
                        Sdata(ctr).TP = 'Not found';
                        ctr=ctr+1;
                        continue;
                    end
                end
                %% DIVIDE DATE RANGE IN ALLOWED SEGMENTS
                [stDates,endDates,stDatesp,endDatesp] = download_segmentation(dStruct,flag1,indx);
                
                %% CHECK DATUM AVAILABILITY
                [datum, datum_p] = datum_selector(dStruct,requested_datum);
                
                %% CHECK TIDAL PREDICTIONS INTERVALS
                [interval,flag3] = prediction_interval_selector(dStruct,flag1, dStruct.greatlakes);
                
                %% ASSIGN INFO TO Sdata
                % Selected Datum For Measurements
                Sdata(ctr).WL_datum = datum;
                % Selected Datum For Predictions
                Sdata(ctr).TP_datum = datum_p;
                % Selected Product For Measurement
                if strcmp(flag1,'6')==1
                    Sdata(ctr).WL_downloaded_product = '6 minutes';
                elseif  strcmp(flag1,'1')==1
                    Sdata(ctr).WL_downloaded_product = 'hourly';
                elseif strcmp(flag1,'6p')==1
                    Sdata(ctr).WL_downloaded_product = '6 minutes preliminary';
                elseif strcmp(flag1,'hilo')==1
                    Sdata(ctr).WL_downloaded_product = 'High/Low';
                else
                    Sdata(ctr).WL_downloaded_product = 'monthly';
                end
                % Slected Product For Predictions
                Sdata(ctr).TP_downloaded_product = interval;
                % record Length
                Sdata(ctr).record_length = dStruct.record_length(indx);
                
                %% NAVD88 ADJUSTMENT
                if strcmp(datum,'NAVD88')
                    datum = 'NAVD';
                end
                if strcmp(datum_p,'NAVD88')
                    datum_p = 'NAVD';
                end
                
                %% BUILD URL & DOWNLOAD MEASURED DATA
                [Sdata] = WL_download(ctr,Sdata,datum,station,timezone,units,format,stDates,endDates,gen_url,options,flag1);
                
                %% DOWNLOAD TIDAL PREDICTIONS
                [Sdata] = tidal_predictions_downloader(ctr,Sdata,datum_p,station,timezone,units,format,interval,stDatesp,endDatesp,gen_url,options,dStruct.greatlakes,flag1,dStruct);
                
                %% MAKE SURE TIME VECTORS ARE THE SAME LENGTH
                if dStruct.greatlakes == 0
                    [Sdata] = vector_length_check(ctr,Sdata,flag1,dStruct.greatlakes);
                end
                %% COMPUTE TOTAL RECORD LENGTH (Non NaN Data)
                [Sdata] = record_length_calc(ctr,Sdata,flag1);
                
                %% ADD DateRange
                Sdata(ctr).Beg = datestr(Sdata(ctr).WL.DateTime(1),'yyyy-mm-dd HH:MM');
                Sdata(ctr).End = datestr(Sdata(ctr).WL.DateTime(end),'yyyy-mm-dd HH:MM');
                
                %% INCREMENT COUNTER FOR PRODUCT LOOP
                ctr = ctr + 1;
            end
        end
    end
    
    %% SAVE DATA
    % save('NOAA_Tidal_Gauge_Data.mat','Sdata');
    
    %% END TIMER
    timing2(1).t2 = datetime('now');
    timing2(1).dt = timing2(1).t2-timing2(1).t1;
    timing2(1).runtime = char(timing2(1).t2-timing2(1).t1);
end

File: noaa-requests/noaa-py/noaapy/wl_downloader_v2.py
# noaa-requests/noaa-py/noaapy/wl_downloader_v2.py
import datetime
import numpy as np
import noaapy


def wl_downloader_v2(
    id_list, station_list, requested_datum, prod, op_mode, d_beg, d_end
):
    """Entry point to the download"""
    if op_mode not in [1, 2]:
        raise ValueError(
            "Please use a valid operational mode (1 - full record, 2 - Specific Date)"
        )

    # Start Timer
    start_time = datetime.datetime.now()

    # Define Initial Variables
    datum = "MSL"  # Preferred Datum for Data Download
    timezone = "GMT"  # Preferred Time Zone for Data Download
    units = "metric"  # Preferred Units for Data Download
    data_format = "csv"  # Download Data Format
    gen_url = "https://api.tidesandcurrents.noaa.gov"  # NOAA CO-OPS URL
    timeout = 800

    # Check if stations are in inventory
    logical_matrix = np.array([[s["id"] == id for id in id_list] for s in station_list])
    max_logical = np.max(logical_matrix, axis=0)
    valid_indices = np.argmax(logical_matrix, axis=0)

    # Single out stations that were not found in StationList
    not_found = [id_list[i] for i, val in enumerate(max_logical) if not val]

    # Create valid station id list
    id_list = [
        station_list[valid_indices[i]]["id"] for i, val in enumerate(max_logical) if val
    ]

    # Initialize Counter
    ctr = 0

    # Initialize Results
    s_data = []

    # Download Data for Each Station
    for i, station_id in enumerate(id_list):
        print(f"------------------ {i + 1} / {len(id_list)} ---------------------")
        for product in prod:
            # Extract Station from Data Structure
            d_struct = next(s for s in station_list if s["id"] == station_id)
            s_data_entry = {
                "id": station_id,
                "name": d_struct["name"],
                "lon": d_struct["lon"],
                "lat": d_struct["lat"],
                "state": d_struct["state"],
            }

            # Check WL Measurements Products Available
            flag4, flag1, indx = noaapy.wl_measurements_product_selector_v2(
                d_struct, product
            )

            if flag4 == 1:
                s_data_entry.update(
                    {
                        "WL_datum": "Not found",
                        "TP_datum": "Not found",
                        "WL_downloaded_product": "Not found",
                        "TP_downloaded_product": "Not found",
                        "record_length": "Not found",
                        "WL": "Not found",
                        "TP": "Not found",
                    }
                )
                s_data.append(s_data_entry)
                ctr += 1
                continue

            # Check if Date Range of Interest is Available
            if op_mode == 2:
                indx, d_end, d_beg, dummy3 = noaapy.date_search(
                    d_struct, d_beg, d_end, indx
                )
                if not dummy3:
                    d_struct["start_date"][indx] = (
                        f"{d_beg.strftime('%Y-%m-%d %H:%M:%S')} GMT"
                    )
                    d_struct["end_date"][indx] = (
                        f"{d_end.strftime('%Y-%m-%d %H:%M:%S')} GMT"
                    )
                else:
                    s_data_entry.update(
                        {
                            "WL_datum": "Not found",
                            "TP_datum": "Not found",
                            "WL_downloaded_product": "Not found",
                            "TP_downloaded_product": "Not found",
                            "record_length": "Not found",
                            "WL": "Not found",
                            "TP": "Not found",
                        }
                    )
                    s_data.append(s_data_entry)
                    ctr += 1
                    continue

            # Divide Date Range in Allowed Segments
            st_dates, end_dates, st_dates_p, end_dates_p = noaapy.download_segmentation(
                d_struct, flag1, indx
            )

            # Check Datum Availability
            datum, datum_p = noaapy.datum_selector(d_struct, requested_datum)

            # Check Tidal Predictions Intervals
            interval, _ = noaapy.prediction_interval_selector(
                d_struct, flag1, d_struct["greatlakes"]
            )

            # Assign Info to Sdata
            s_data_entry.update(
                {
                    "WL_datum": datum,
                    "TP_datum": datum_p,
                    "WL_downloaded_product": product_label(flag1),
                    "TP_downloaded_product": interval,
                    "record_length": d_struct["record_length"][indx],
                }
            )

            # NAVD88 Adjustment
            if datum == "NAVD88":
                datum = "NAVD"
            if datum_p == "NAVD88":
                datum_p = "NAVD"

            # Build URL & Download Measured Data
            s_data = noaapy.wl_download(
                ctr,
                s_data,
                datum,
                station_id,
                timezone,
                units,
                data_format,
                st_dates,
                end_dates,
                gen_url,
                timeout,
                flag1,
            )

            # Download Tidal Predictions
            s_data = tidal_predictions_downloader(
                ctr,
                s_data,
                datum_p,
                station_id,
                timezone,
                units,
                data_format,
                interval,
                st_dates_p,
                end_dates_p,
                gen_url,
                timeout,
                d_struct["greatlakes"],
                flag1,
                d_struct,
            )

            # Make Sure Time Vectors Are the Same Length
            if d_struct["greatlakes"] == 0:
                s_data = vector_length_check(ctr, s_data, flag1, d_struct["greatlakes"])

            # Compute Total Record Length (Non NaN Data)
            s_data = record_length_calc(ctr, s_data, flag1)

            # Add DateRange
            s_data_entry.update(
                {
                    "Beg": s_data[ctr]["WL"]["DateTime"][0].strftime("%Y-%m-%d %H:%M"),
                    "End": s_data[ctr]["WL"]["DateTime"][-1].strftime("%Y-%m-%d %H:%M"),
                }
            )

            # Increment Counter for Product Loop
            ctr += 1

    # End Timer
    end_time = datetime.datetime.now()
    run_time = end_time - start_time

    return s_data, not_found


def product_label(flag):
    labels = {
        "6": "6 minutes",
        "1": "hourly",
        "6p": "6 minutes preliminary",
        "hilo": "High/Low",
        "m": "monthly",
    }
    return labels.get(flag, "unknown")
